File: client/test/link_code_test.dart
import 'package:elum/features/link/domain/link_code.dart';
import 'package:flutter_test/flutter_test.dart';

/// 연결 암호 다루기 (이슈 #205 · 명세 §5-0).
///
/// 서버와 **같은 집합**을 알고 있어야 한다. 더 넓게 받으면 서버가 404를 돌려주고,
/// 좁게 받으면 맞는 암호를 우리가 먼저 막는다.
void main() {
  test('서버와 같은 30자 — 헷갈리는 글자는 없다', () {
    expect(LinkCode.alphabet.length, 30);
    for (final banned in '0O1ILU'.split('')) {
      expect(LinkCode.alphabet, isNot(contains(banned)),
          reason: '$banned 는 불러줄 때 헷갈려 서버가 만들지 않는다');
    }
  });

  group('화면 표시', () {
    test('3-3으로 끊어 보여준다 — 여섯을 붙이면 불러주다 자리를 놓친다', () {
      expect(LinkCode.grouped('A7K3M9'), 'A7K 3M9');
    });

    test('여섯 자가 아니면 손대지 않는다', () {
      expect(LinkCode.grouped('A7K'), 'A7K');
    });
  });

  group('입력 정규화', () {
    test('소문자로 쳐도 된다', () {
      expect(LinkCode.normalize('a7k3m9'), 'A7K3M9');
    });

    test('화면에 보이는 대로 띄어 써도 된다', () {
      expect(LinkCode.normalize('a7k 3m9'), 'A7K3M9');
      expect(LinkCode.normalize('A7K-3M9'), 'A7K3M9');
    });
  });

  group('보내기 전 거르기', () {
    test('맞는 모양', () {
      expect(LinkCode.hasValidShape('A7K3M9'), isTrue);
    });

    test('길이가 다르면 보내지 않는다', () {
      expect(LinkCode.hasValidShape('A7K3M'), isFalse);
      expect(LinkCode.hasValidShape('A7K3M99'), isFalse);
    });

    test('만들 수 없는 글자면 보내지 않는다 — 시도 횟수만 축낸다', () {
      expect(LinkCode.hasValidShape('A0K3M9'), isFalse);
      expect(LinkCode.hasValidShape('AIK3M9'), isFalse);
    });
  });
}

File: server/src/test/java/com/chuseok22/elumserver/link/core/LinkCodeTest.java
package com.chuseok22.elumserver.link.core;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.stream.IntStream;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

/**
 * 연결 암호 규칙 (이슈 #200 · 명세 §5-0).
 *
 * <p>보호자가 불러주고 이룸이가 받아적는다. 헷갈리는 글자가 하나라도 섞이면
 * 그 자리에서 막히므로, "웬만하면 안 나온다"가 아니라 **한 번도 안 나온다**를 고정한다.
 */
class LinkCodeTest {

  @Test
  @DisplayName("헷갈리는 글자(0 O 1 I L U)는 만들지 않는다")
  void generate_neverContainsConfusingCharacters() {
    String banned = "0O1ILU";

    IntStream.range(0, 3000).forEach(i -> {
      String code = LinkCode.generate();
      assertThat(code).hasSize(6);
      for (char c : banned.toCharArray()) {
        assertThat(code)
          .withFailMessage("%d번째 생성에서 금지 문자 '%c'가 나왔습니다: %s", i, c, code)
          .doesNotContain(String.valueOf(c));
      }
    });
  }

  @Test
  @DisplayName("소문자로 쳐도 대문자로 받는다")
  void normalize_upperCases() {
    assertThat(LinkCode.normalize("a7k3m9")).isEqualTo("A7K3M9");
  }

  @Test
  @DisplayName("화면이 3-3으로 보여주므로 사이 공백을 받아준다")
  void normalize_stripsSpacing() {
    assertThat(LinkCode.normalize("A7K 3M9")).isEqualTo("A7K3M9");
    assertThat(LinkCode.normalize("A7K-3M9")).isEqualTo("A7K3M9");
  }

  @Test
  @DisplayName("null 이 와도 터지지 않는다")
  void normalize_nullSafe() {
    assertThat(LinkCode.normalize(null)).isEmpty();
  }

  @Test
  @DisplayName("우리가 만들 수 없는 모양은 걸러낸다")
  void hasValidShape() {
    assertThat(LinkCode.hasValidShape("A7K3M9")).isTrue();
    assertThat(LinkCode.hasValidShape("A7K3M")).isFalse();     // 5자리
    assertThat(LinkCode.hasValidShape("A7K3M99")).isFalse();   // 7자리
    assertThat(LinkCode.hasValidShape("A0K3M9")).isFalse();    // 금지 문자 0
    assertThat(LinkCode.hasValidShape("A7K3M!")).isFalse();    // 특수문자
  }

  @Test
  @DisplayName("알파벳은 30자다 — 6자리로 7억 가지")
  void alphabetSize() {
    assertThat(LinkCode.ALPHABET).hasSize(30);
    assertThat(LinkCode.ALPHABET.chars().distinct().count()).isEqualTo(30);
  }
}
